controllers/grid/queries/form/QueryForm.inc.php

```diff
@@ -323,6 +323,88 @@ function readInputData() {
+			$userGroupDao = DAORegistry::getDAO('UserGroupDAO'); /* @var $userGroupDao UserGroupDAO */
+			$assignedRoles = [];
+			$usersAssignments = $stageAssignmentDao->getBySubmissionAndStageId($submissionId, $query->getStageId(), null, $user->getId());
+			while ($usersAssignment = $usersAssignments->next()) {
+				$userGroup = $userGroupDao->getById($usersAssignment->getUserGroupId());
+				$assignedRoles[] = $userGroup->getRoleId();
+			}
+
+			// if current user is editor, forbid several blind participants (reviewers and author) in a review discussion
+			if ($user->hasRole([ROLE_ID_SITE_ADMIN], CONTEXT_SITE) ||
+				array_intersect([ROLE_ID_MANAGER, ROLE_ID_SUB_EDITOR], $assignedRoles)) {
+
+					if ($query->getStageId() == WORKFLOW_STAGE_ID_EXTERNAL_REVIEW || $query->getStageId() == WORKFLOW_STAGE_ID_INTERNAL_REVIEW) {
+						// get authors assigned to the submission (also be provided in the participant selection list)
+						$authorIds = [];
+						$authorAssignments = $stageAssignmentDao->getBySubmissionAndRoleId($submissionId, ROLE_ID_AUTHOR);
+						while ($authorAssignment = $authorAssignments->next()) {
+							if (!in_array($authorAssignment->getUserId(), $authorIds)) {
+								$authorIds[] = $authorAssignment->getUserId();
+							}
+						}
+
+						$reviewAssignmentDao = DAORegistry::getDAO('ReviewAssignmentDAO'); /* @var $reviewAssignmentDao ReviewAssignmentDAO */
+						// get the selected participants
+						$newParticipantIds = $this->getData('users');
+						$blindReviewerExists = $authorExists = $displayError = false;
+						foreach ($newParticipantIds as $participantId) {
+							if (in_array($participantId, $authorIds)) {
+								if ($blindReviewerExists) {
+									$displayError = true;
+									break;
+								} else {
+									$authorExists = true;
+								}
+							} else {
+								// get all review assignments for the participant for the submission
+								// if one assignemnt is blind, then the participant should stay anonymous in all review discussions
+								// because the review discussions are independent of the review rounds
+								$reviewAssignments = $reviewAssignmentDao->getBySubmissionReviewer($submissionId, $participantId);
+								foreach($reviewAssignments as $reviewAssignment) {
+									if ($reviewAssignment->getReviewMethod() != SUBMISSION_REVIEW_METHOD_OPEN) {
+										// this is a blind review assignment
+										// display the erros if a blind participant or an author is already selected
+										if ($blindReviewerExists || $authorExists) {
+											$displayError = true;
+											break 2;
+										} else {
+											$blindReviewerExists = true;
+											break;
+										}
+									}
+								}
+							}
+						}
+						if ($displayError) {
+							$this->addError('users', __('editor.review.discussion.errorAnonymousParticipants'));
+							$this->addErrorField('users');
+						}
```

Comment:
Can I suggest a small adjustment that might take less logic? First, get a count of all authors and anonymous reviewers, then apply the validation logic. Like this:

```php
$participantIds = $this->getData('users');
$authorAssignments = $stageAssignmentDao->getBySubmissionAndRoleId($submissionId, ROLE_ID_AUTHOR);
$anonymousCount = 0;
while ($authorAssignment = $authorAssignments->next()) {
	if ($authorAssignment->getRoleId() == ROLE_ID_AUTHOR && in_array($authorAssignment->getUserId(), $participantIds) {
		$anonymousCount++;
	}
}
$reviewAssignments = $reviewAssignmentDao->getBySubmissionId($submissionId);
foreach ($reviewAssignments as $reviewAssignment) {
	if (in_array($authorAssignment->getUserId(), $participantIds)
			&& in_array($reviewAssignment->getReviewMethod(), [SUBMISSION_REVIEW_METHOD_ANONYMOUS, SUBMISSION_REVIEW_METHOD_DOUBLEANONYMOUS])) {
		$anonymousCount++;
	}
}
if ($anonymousCount > 1) {
	// apply error
}
```

Reply:
Here I am only not sure if several authors are selected and no reviewer -- they could see each other, or?
I would then also need to consider other roles (if this should apply regardless of stage and if stage participants are configurable), so there might be something like this then:
if participant role is not editor and not blind reviewer: count++
if participant is a blind reviewer: anonymousCount++
if (count > 0 && anonymousCount > 0) or (anonymousCount > 1): display error
I will then still have to see to do it with less logic, as you said... :-)
What do you think?

I will also have to see how to retrieve the participant roles...

Reply:
> Here I am only not sure if several authors are selected and no reviewer -- they could see each other, or?

Ack! You're right! Either keep separate counts of the authors and reviewers. Or just break the `while` loop after the first author is found. That way it will only increment once if an author is in there.

> I would then also need to consider other roles

Ok, yeah, I guess if someone has an assistant or subeditor assignment to another stage but not the review stage, they shouldn't have access. I'm trying to think whether it would be possible to create such a discussion. The UI won't show them the participant IDs for selection when creating a discussion, but is it possible for a malicious user to add a participant who is not assigned to the stage? It seems like we can maybe tackle this problem from that direction: it shouldn't be possible to create a discussion in X stage with a participant who is not assigned to that stage.